### PR TITLE
DAOS-16329 chk: maintenance mode after checking pool with dryrun - b26

### PR DESCRIPTION
**Pull request by Nasf-Fan**
Sometimes, after system shutdown unexpectedly, the users may expect to check their critical data under some kind of maintenance mode. Under such mode, no user data can be modified or moved or aggregated. That will guarantee no further potential (DAOS logic caused) damage can happen during the check.

For such purpose, we will enhance current DAOS CR logic with --dryrun option to allow the pool (after check) to be opened as immutable with disabling some mechanism that may potentially cause data modification or movement (such as rebuild or aggregation).

Under such mode, if client wants to connect to the pool, the read-only option must be specified. Similarly for opening container in such pool.

Test-tag: pr cat_recov
Allow-unstable-test: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/cart/util/cart_logtest.py
Comment (daltonbohning) [inline src/tests/ftest/cart/util/cart_logtest.py:212]: Why are we ignoring cont_open here instead of fixing the errors? And what are the errors we're ignoring?

**Review by daltonbohning — DISMISSED** [review]

File: src/tests/ftest/cart/util/cart_logtest.py
Comment (Nasf-Fan) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: This is some temporary workaround for the following NLT warning:
https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-14912/10/VM_test/folder.-412761578/

`cont_open() 07139793/8d9ab078/6908b833: No permission to open the container with flags 2, capas 7/0 warning in strict mode`

I do not know how the new warning is valuable for us.

Comment (Nasf-Fan) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: I do not know how to fix such NLT warning because even if I modified some existing logic line in `cont_open`, the CI test will also report similar NLT warning. @daltonbohning , any suggestion for that? Thanks!

Comment (daltonbohning) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: Sorry, that was the wrong code snippet. But I think the problem is the same

Comment (Nasf-Fan) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: > I think the problem is it doesn't follow the standard format with the rc in the error. The same would apply for other error messages as well

You mean that we need log the "rc" in the D_ERROR(), right?

Comment (Nasf-Fan) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: @daltonbohning , what is the new NLT warning means:

https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14984/7/VM_test/new/

The warning corresponding to the following changes in the patch (https://github.com/daos-stack/daos/pull/14984):
`
2300         if (((flags & DAOS_COO_EX) && !ds_sec_cont_can_open_ex(sec_capas)) ||
2301             ((flags & DAOS_COO_RW) && !ds_sec_cont_can_modify(sec_capas))) {
2302                 rc = -DER_NO_PERM;
2303                 D_ERROR(DF_UUID "/" DF_UUID "/" DF_UUID ": failed to open the container "
2304                         "with flags " DF_X64 ", capas " DF_X64 "/" DF_X64 ": " DF_RC "\n",
2305                         DP_UUID(cont->c_svc->cs_pool_uuid), DP_UUID(pool_hdl->sph_uuid),
2306                         DP_UUID(cont->c_uuid), flags, pool_hdl->sph_sec_capas, sec_capas,
2307                         DP_RC(rc));
`

Comment (daltonbohning) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: Sorry, I'm not familiar enough with NLT to know what it's complaining about.
Looking at the code, it might be complaining that the error is logged at all?

Comment (Nasf-Fan) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: Honestly, I do not know. The NLT warning is still there even if without my change. But because I added more information in such log line, then it failed related CI test results. Personally, I do not think such warning is valuable for us. That is why I put `cont_open` into the ignore list.

Comment (Nasf-Fan) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: According to the commit edeafc14d87a5d839f4dba482efc7a54cca89592 that introduced the IGNORED_FUNCTIONS list, I did not find any special reason for adding "rdb_timerd" into such list except reducing unnecessary "fail" report.

Comment (Nasf-Fan) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: In fact, such NLT warning for `cont_open` log exists on current master, it is not reported as CI failure if no patch touch related code. But my patch touches it because I need more information in the log, then it failed the CI test.

From the test logic perspective (although I do not agree), it seems that such log line should not be there.
@daltonbohning , do you still think it is not an acceptable solution to "skip" the existing `cont_open` NLT warning?

Comment (Nasf-Fan) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: Here is an example for the NLT warning existence:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15055/1/VM_test/

Comment (daltonbohning) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: I'm not very familiar with NLT so I don't know what is best here. It just seems bad to disable a subset of testing because it's failing. @ashleypittman is the best source here but is on sabbatical.
If it's pre-existing, maybe this applies? Copied from our dev channel

>Tips from Ashley:
the way Jenkins works is it checks if messages are the same, and obviously it can't just check the line number for this as that changes.  It takes a snapshot of the 5 lines before and after the line where the error is reported, and if these are the same then the error is treated as the same.  If these lines are modified then the error does not match and therefore you get "one removed error", "one error added"
>
>in which case the procedure is the same as for adding a new error, complete your PR as you want it, run with Allow-unstable-test: true commit tags, the PR will need force-landing but once it's landed master will run and create new reference builds so other PRs should not be affected.

I.e. add `Allow-unstable-test: true` to your commit pragmas and request force landing, since it's a pre-existing error

Comment (Nasf-Fan) [reply src/tests/ftest/cart/util/cart_logtest.py:212]: let me try.

**Review by daltonbohning — COMMENTED** [review]
Will reserve approval for those familiar with core code, I think we will need to force land for the existing NLT failure

**Comment by Nasf-Fan** [comment]
Ping reviewers for the backport patch. Thanks!

**Review by kjacque — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Comment by daltonbohning** [comment]
FYI - will need to wait for 2.6.1 GA, then merge approval for this ticket, then we can merge.

**Comment by daltonbohning** [comment]
Ticket does not have approval to merge to 2.6.2

**Comment by Nasf-Fan** [comment]
> Ticket does not have approval to merge to 2.6.2

So we will not land this patch to any 2.6 based branch any longer, right?

**Comment by daltonbohning** [comment]
> > Ticket does not have approval to merge to 2.6.2
> 
> So we will not land this patch to any 2.6 based branch any longer, right?

It means we need to go through the merge approval process for this ticket/PR

**Comment by Nasf-Fan** [comment]
> Ticket does not have approval to merge to 2.6.2

It is shown on the Jira ticket DAOS-16329 that `Approved to Merge for 2.6.2 Community Release`

**Comment by daltonbohning** [comment]
> > Ticket does not have approval to merge to 2.6.2
> 
> It is shown on the Jira ticket DAOS-16329 that `Approved to Merge for 2.6.2 Community Release`

To be clear, I posted this on Oct 8 and `Approved to Merge for 2.6.2 Community Release` was not set until Oct 9. We need approval to merge _**before**_ requesting gatekeeper

**Pull request merged**
